lib/rules/no-shared-component-data.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * @fileoverview When using the data property on a component (i.e. anywhere except on new Vue), the value must be a function that returns an object.
+ * @author Armano
+ */
+'use strict'
+
+const utils = require('../utils')
+
+function create (context) {
+  // ----------------------------------------------------------------------
+  // Public
+  // ----------------------------------------------------------------------
+
+  return Object.assign({},
+    utils.executeOnVueComponent(context, (obj) => {
+      obj.properties
+        .filter(p =>
+          p.type === 'Property' &&
+          p.key.type === 'Identifier' &&
+          p.key.name === 'data' &&
+          p.value.type !== 'FunctionExpression' &&
+          p.value.type !== 'Identifier'
+        ).forEach(cp => {
+          context.report({
+            node: cp.value,
+            message: 'Component `data` must be a function.'
```

Comment:
What do you think about changing the message to:
`data` property in component must be a function